pkg/arvo/sys/hoon.hoon

```diff
@@ -11534,114 +11557,175 @@
     ==
   ++  docs
     |%
+    ::  +apex: prefix comment. may contain batch comments.
     ::
-    ::  above core
-    ::
+    ::    when a prefix doccord is parsed, it is possible that there is no +gap
+    ::    afterward to be consumed, so we add an additional newline and set back
+    ::    the line of the n
     ++  apex
+      %+  knee  *whit  |.  ~+
       ;~  plug
-        =/  ron  (punt (indo noel))
-        (punt (ifix [ron ron] (into head)))             ::  label
-      ::
-        =/  ron  (punt (indo null))
-        (ifix [ron ron] (punt body))                    ::  body
+        |=  tub=nail
+        =/  vex
+          %.  tub
+          %-  star  %-  myth
+          ;~(pfix (punt leap) (star ace) col col ;~(pose larg smol))
+        ?~  q.vex  vex
+        :-  p=p.vex
+        %-  some
+        ?~  p.u.q.vex
+          [p=~ q=q.u.q.vex]
+        :-  p=(malt p.u.q.vex)
+        q=`nail`[[(dec p.p.q.u.q.vex) q.p.q.u.q.vex] ['\0a' q.q.u.q.vex]]
       ::
-        (cook malt (star fill))                         ::  definitions
-        (easy ~)                                        ::  defs used (none)
+        (easy ~)
       ==
     ::
-    ::  backward line
-    ::
+    ::  +apse: postfix comment.
     ++  apse
+      %+  knee  *help  |.  ~+
       ;~  pose
-        %+  cook  |=([a=term b=cord] %*(. *whit def (my [a b ~] ~)))
-        (exit fine)
-      ::
-        %+  cook  |=(a=cord %*(. *whit boy `[a ~]))
+        %+  cook  |=(a=cord %*(. *help summary.crib a))
+        ::TODO: if there is a $link then interpret it as the start
+        ::of a prefix comment instead of postfix
+        ::;~(less (exit ;~(plug (plus en-link) col ace)) (exit line))
         (exit line)
       ::
-        (easy *whit)
+        (easy *help)
       ==
     ::
-    ::
-    ++  beer
-      |=  $:  lab=(unit term)
-              boy=(unit (pair cord (list sect)))
-              def=(list (pair (pair term cord) (list sect)))
+    ++  leap                                            ::  whitespace less docs
+      %+  cold  ~
+      %-  plus
+      ;~  plug
+        (star ace)
+        ;~  pose
+          ;~(plug (just '\0a') (star ace))
+          ;~  plug  col  col
+            ;~(less ;~(pose larg smol) ;~(plug (star prn) (just '\0a')))
           ==
-      ^-  whit
-      =;  def  [lab boy (malt def) ~]
-      (turn def |=([[a=term b=cord] c=(list sect)] [a [b c]]))
-    ::
-    ::
-    ++  body
-      ;~  pose
-        ;~  plug                                        :: can duplicate ::
-          (into ;~(pfix (punt ;~(plug null col gar step)) line))
-          (easy ~)
         ==
+      ==
+    ::
+    ::  +smol: 2 aces then summary, 4 aces then paragraphs.
+    ++  smol
+      ;~  pfix
+        step
         ;~  plug
-          (into ;~(pfix step line))
-          (rant text)
+          ;~  plug
+            (plus en-link)
+            ;~(sfix ;~(pfix col ace (cook crip (star prn))) (just '\0a'))
+          ==
```

Comment:
I chose to do this style of comment this way because otherwise every comment that starts with `::` then two aces will become a doccord, and I wanted to leave some wiggle room for people to write ordinary comments that don't end up in doccords, since it might not necessarily make sense for them to be attached to whatever they're attaching to. This probably filters out some comments for which this is not the case, but will definitely eliminate a lot of noise that would negatively affect the usability of doccords.

I would estimate based purely on overall impression in the code that comments already written in the batch comment style with a `$link` are useful and accurate, rather than things like notes to the programmer reminding themselves of unfinished work attached to a random atom.

So its sort of a spam prevention measure for usability with the existing codebase, and a way for people who simply don't want to use doccords to write comments like they normally do and avoid it entirely. I don't know whether or not that's the principled way to do this but atm this feels like more of a question of taste than something that really has a right or wrong answer.


For `+larg`, we actually do parse as an empty link, and this signifies to the parser to just attach the `%note` to the following `hoon` or `spec`.


Needing a `col ace` prefix is a mistake I made rewriting the parser, I'll fix that.

Reply:
https://github.com/urbit/urbit/pull/5873/commits/ca3140a8b3997dd08976f8c09bfd1bdeebd4b14e

Reply:
See also https://github.com/urbit/urbit/pull/5873/commits/627827f02959b5ddb991799b1e1384467690a0be